dali/kernels/imgproc/pointwise/linear_transformation_gpu_test.cu

```diff
@@ -0,0 +1,244 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <vector>
+#include <tuple>
+#include "dali/core/geom/mat.h"
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/tensor_shape.h"
+#include "dali/kernels/common/copy.h"
+#include "dali/kernels/test/tensor_test_utils.h"
+#include "dali/kernels/test/kernel_test_utils.h"
+#include "dali/kernels/imgproc/pointwise/linear_transformation_gpu.h"
+#include "dali/kernels/imgproc/color_manipulation/color_manipulation_test_utils.h"
+
+namespace dali {
+namespace kernels {
+namespace test {
+
+
+namespace {
+
+/**
+ * Rounding to nearest even (like GPU does it)
+ */
+template <class Out>
+std::enable_if_t<std::is_integral<Out>::value, Out> custom_round(float val) {
+  return ConvertSat<Out>(std::nearbyint(val));
+}
+
+
+template <class Out>
+std::enable_if_t<!std::is_integral<Out>::value, Out> custom_round(float val) {
+  return val;
+}
+
+
+constexpr int kNDims = 3;
+constexpr int kNChannelsIn = 5;
+constexpr int kNChannelsOut = 2;
+
+
+}  // namespace
+
+template <class InputOutputTypes>
+class LinearTransformationGpuTest : public ::testing::Test {
+  using In = typename InputOutputTypes::In;
+  using Out = typename InputOutputTypes::Out;
+
+ public:
+  LinearTransformationGpuTest() {
+    input_host_.resize(dataset_size(in_shapes_));
+  }
+
+
+  void SetUp() final {
+    std::mt19937_64 rng;
+    UniformRandomFill(input_host_, rng, 0., 10.);
+    calc_output();
+    CUDA_CALL(cudaMalloc(&input_device_, sizeof(In) * dataset_size(in_shapes_)));
+    CUDA_CALL(cudaMemcpy(input_device_, input_host_.data(), input_host_.size() * sizeof(In),
+                         cudaMemcpyDefault));
+    CUDA_CALL(cudaMalloc(&output_, dataset_size(out_shapes_) * sizeof(Out)));
+    cudaDeviceSynchronize();
+  }
+
+
+  In *input_device_;
+  Out *output_;
+  std::vector<In> input_host_;
+  std::vector<Out> ref_output_;
+  std::vector<TensorShape<kNDims>> in_shapes_ = {{4, 3, kNChannelsIn}};
+  std::vector<TensorShape<kNDims>> out_shapes_ = {{4, 3, kNChannelsOut}};
+  mat<kNChannelsOut, kNChannelsIn, float> mat_{{{1, 2, 3, 4, 5}, {6, 7, 8, 9, 10}}};
+  vec<kNChannelsOut, float> vec_{42, 69};
+  std::vector<mat<kNChannelsOut, kNChannelsIn, float>> vmat_ = {mat_};
+  std::vector<vec<kNChannelsOut, float>> vvec_ = {vec_};
+  std::vector<Roi<2>> rois_ = {{{1, 1}, {2, 2}}};
+
+
+  void calc_output() {
+    for (size_t i = 0; i < input_host_.size(); i += kNChannelsIn) {
+      for (size_t j = 0; j < kNChannelsOut; j++) {
+        float res = vec_.v[j];
+        for (size_t k = 0; k < kNChannelsIn; k++) {
+          res += static_cast<float>(input_host_[i + k]) * mat_.at(j, k);
+        }
+        ref_output_.push_back(custom_round<Out>(res));
```

Comment:
All your coefficients are integer - why do you need precise rounding?

Reply:
Removed